CON-461: Add reference for Exports operations (#504)
MewsSystems/gitbook-connector-api: operations/exports.md

# Exports

> ### Restricted!
> These operations are currently in beta-test and as such are subject to change.

## Get all exports

> Get exports created by any integration for the enterprise associated with the provided access token. Only exports with status of `Pending`, `Processing`, and `Success` are returned, unless the exports are retrieved using the `ExportIds` property. Note this operation uses [Pagination](../guidelines/pagination.md) and supports [Portfolio Access Tokens](../guidelines/multi-property.md).

Author: I agree that'd be a cleaner design, although we don't expose the status filter yet.

But honestly, I'm not even sure what is use case for accessing exports without specifying their ID other than maintaining semantics of `getAll` endpoint. The initial design talked about polling with explicit export ID, and I'm very worried about expanding filtering features on this endpoint, at least before learning about relevant use case.

> Get exports created by any integration for the enterprise associated with the provided access token. Only exports with status of `Pending`, `Processing`, and `Success` are returned, unless the exports are retrieved using the `ExportIds` property. Note this operation uses [Pagination](../guidelines/pagination.md) and supports [Portfolio Access Tokens](../guidelines/multi-property.md).

Author: As discussed in the meeting, initially we will allow only querying by export IDs without listing. We can expand the scope of this endpoint once we gather feedback from beta users.

| `ExpiresUtc` | string | optional | Expiration date and time of the export in UTC timezone in ISO 8601 format. After this time the [exported files](#exported-file) may no longer be available for download. |

#### Export status

* `Pending` - export was created and didn't start yet
* `Processing` - export process has started
* `Success` - export has been successfully finished and [exported files](#exported-file) are available for download
* `Expired` - export has expired and exported files are no longer available
* `Failed` - export has failed and exported files are not available; you can try creating a new export

#### Exported file

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `Url` | string | required | URL of the exported file for download in [JSON Lines](https://jsonlines.org/) format. |
| `SizeInBytes` | number | required | Total size of the exported file in bytes. |

## Add export

Create a pending export. Export will include all entities of the specified `EntityType` from all enterprises accessible for the integration (if [Portfolio Access Tokens](../guidelines/multi-property.md) is used).

> If there is already an export with status `Pending` or `Processing`, request for new export will be rejected.

Reviewer: ```suggestion
If there is already an export with status `Pending` or `Processing` for the specified entity type, request for new export will be rejected.
```

> # Exports

Reviewer 2: I think this should have a use case described as well for sure 👍 Or we can make it later, when we remove restricted.

Author: I believe use case should go on a separate page, let's keep the reference for reference only.

> ### Restricted!
> These operations are currently in beta-test and as such are subject to change.

## Get all exports

Get exports by provided `ExportIds`. Note this operation supports [Portfolio Access Tokens](../guidelines/multi-property.md).

### Request

`[PlatformAddress]/api/connector/v1/exports/getAll`

> ```json

Reviewer 2: Remove json. We don't do that, we're not sure if there wasn't any issues with it 🤔 that must by checked by mike. Do it everywhere.

Author: Yeah, I saw there's `javascript` instead. Probably should check whether GitBook can crunch it.

{
    "ClientToken": "E0D439EE522F44368DC78E1BFB03710C-D24FB11DBE31D4621C4817E028D9E1D",
    "AccessToken": "C66EF7B239D24632943D115EDE9CB810-EA00F8FD8294692C940F6B5A8F9453D",
    "Client": "Sample Client 1.0.0",
    "ExportIds": [
        "3fa85f64-5717-4562-b3fd-2c963f66afa6",
        "09708665-0e31-4b23-b337-b0a000be0df0",
        "706dc6d5-9511-4751-825e-538ce99da2ce",
        "f776f20a-6f1a-4ddf-93f4-9dae95261415"

>     ]

Reviewer 2: I think you are missing `Limitation` in the example? It's required

Author: No, per the last discussion it was removed since don't allow listing on this endpoint. See also CON-648.

Reviewer 2: If that so, remove it from the list of properties below, where it say required 👍

>     ]

Author: Oh, this was also removed since it doesn't make sense to have it for pulling by IDs. I will remove the parameter from the table below.

}
```

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `ClientToken` | string | required | Token identifying the client application. |
| `AccessToken` | string | required | Access token of the client application. |
| `Client` | string | required | Name and version of the client application. |

> | `ExportIds` | array of string | optional, max 1000 items | Unique identifiers of the [Exports](#export). |

Reviewer 2: ```suggestion
| `ExportIds` | array of string | required, max 1000 items | Unique identifiers of the [Exports](#export). |
```
This will be required right?

Author: Dammit, right! 😅

> | `ExportIds` | array of string | optional, max 1000 items | Unique identifiers of the exports. If not specified, the operation returns exports with Status of `Pending`, `Processing`, and `Success` for the enterprise associated with the Access Token. |

Reviewer 2: ```suggestion
| `ExportIds` | array of string | optional, max 1000 items | Unique identifiers of the [Exports](#export). If not specified, the operation returns exports with Status of `Pending`, `Processing`, and `Success` for the enterprise associated with the Access Token. |
```

### Response

```json
{
    "Exports": [
        {
            "Id": "3fa85f64-5717-4562-b3fd-2c963f66afa6",
            "Status": "Success",
            "EntityType": "OrderItem",
            "ExpiresUtc": "2023-10-26T11:42:28Z",
            "Files": [
                {
                    "Url": "https://example.com/exports/3fa85f64-5717-4562-b3fd-2c963f66afa6-1.jsonl?example=signature",
                    "SizeInBytes": 1215279
                },
                {
                    "Url": "https://example.com/exports/3fa85f64-5717-4562-b3fd-2c963f66afa6-2.jsonl?example=signature",
                    "SizeInBytes": 1398362
                },
            ]
        },
        {
            "Id": "09708665-0e31-4b23-b337-b0a000be0df0",
            "Status": "Pending",
            "EntityType": "OrderItem",
            "Files": [],
            "ExpiresUtc": null
        },
        {
            "Id": "706dc6d5-9511-4751-825e-538ce99da2ce",
            "Status": "Processing",
            "EntityType": "OrderItem",
            "Files": [],
            "ExpiresUtc": null
        },
        {

>             "Id": "706dc6d5-9511-4751-825e-538ce99da2ce",

Reviewer 2: ```suggestion
            "Id": "f776f20a-6f1a-4ddf-93f4-9dae95261415",
```
Wrong Id I guess? It's duplicated with the one above.

Author: Right, pasted the wrong one, good catch!

            "Status": "Expired",
            "EntityType": "OrderItem",
            "Files": [],
            "ExpiresUtc": "2023-10-24T14:12:30Z"
        }
    ]
}
```

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |

> | `Exports` | array of [Exports](#export) | required | Requested exports. |

Reviewer 2: ```suggestion
| `Exports` | array of [Export](#export) | required | Requested exports. |
```
Singular in this case

#### Export

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `Id` | string | required | Unique identifier of the export. |
| `Status` | string [Export status](#export-status) | required | Current status of the export. |
| `EntityType` | string [Entity type](#entity-type) | required | Type of exported entities. |

> | `Files` | array of [Exported files](#exported-file) | required | Files with exported data, if available. |

Reviewer 2: ```suggestion
| `Files` | array of [Exported files](#exported-file) | required | Files with exported data. Empty if no files are available. |
```
I would change it little bit because at first reading I was like, why is it required when you are saying "if available" so I was thinking it can be null 😁 ?

> | `ExpiresUtc` | string | optional | Expiration date and time of the export in UTC timezone in ISO 8601 format. After this time the [exported files](#exported-file) may no longer be available for download. |

Reviewer 2: ```suggestion
| `ExpiresUtc` | string | optional | Expiration date and time of the export in UTC timezone in ISO 8601 format. After this time the [Exported files](#exported-file) may no longer be available for download. |
```
Everywhere you mention something, it should start with capital if I'm not mistaken.

#### Export status

* `Pending` - export was created and didn't start yet.
* `Processing` - export process has started.
* `Success` - export has been successfully finished and [Exported files](#exported-file) are available for download.
* `Expired` - export has expired and exported files are no longer available.

> * `Failed` - export has failed and exported files are not available; you can try creating a new export.

Author: Well, there isn't any field for that yet, in fact this state is currently unused in the code. We will likely add a separate string field (or perhaps an object) to describe the error. Suggestions welcome!

> * `Failed` - export has failed and exported files are not available; you can try creating a new export.

Reviewer 3: I don't know what the failure modes are, but it makes sense to me to have a reason string. But sure, this could be added later when you better know the failure modes.

> * `Pending` - export was created and didn't start yet
> * `Processing` - export process has started
> * `Success` - export has been successfully finished and [exported files](#exported-file) are available for download
> * `Expired` - export has expired and exported files are no longer available
> * `Failed` - export has failed and exported files are not available; you can try creating a new export

Reviewer 2: Is there a chance we would add another status? 🤔 If so, we should mark it as extendable.

```suggestion
* `Pending` - export was created and didn't start yet
* `Processing` - export process has started
* `Success` - export has been successfully finished and [exported files](#exported-file) are available for download
* `Expired` - export has expired and exported files are no longer available
* `Failed` - export has failed and exported files are not available; you can try creating a new export
* ...
```

Author: I find it very unlikely we'll need another status (in fact, `Failed` is currently unused) and we're still in beta so we can change it as we desire.

> * `Pending` - export was created and didn't start yet
> * `Processing` - export process has started
> * `Success` - export has been successfully finished and [exported files](#exported-file) are available for download
> * `Expired` - export has expired and exported files are no longer available
> * `Failed` - export has failed and exported files are not available; you can try creating a new export

Reviewer 2: ```suggestion
* `Pending` - export was created and didn't start yet.
* `Processing` - export process has started.
* `Success` - export has been successfully finished and [Exported files](#exported-file) are available for download.
* `Expired` - export has expired and exported files are no longer available.
* `Failed` - export has failed and exported files are not available; you can try creating a new export.
```
Missing dots at the end, I think it's everywhere like that + capital E

#### Exported file

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `Url` | string | required | URL of the exported file for download in [JSON Lines](https://jsonlines.org/) format. |
| `SizeInBytes` | number | required | Total size of the exported file in bytes. |

## Add export

Create a pending export. Export all entities of the specified `EntityType` within the enterprise. Note this operation supports [Portfolio Access Tokens](../guidelines/multi-property.md) but don't have filtering by `EnterpriseIds` or `ChainIds` and will include entities from all enterprises within the scope of the portfolio.

> 

Reviewer 3: Small point, but I prefer one block of text here rather than two.

> Create a pending export. Export will include all entities of the specified `EntityType` from all enterprises accessible for the integration (if [Portfolio Access Tokens](../guidelines/multi-property.md) is used).

Reviewer 2: ```suggestion
Create a pending export. Export all entities of the specified `EntityType` within the enterprise. Note this operation supports [Portfolio Access Tokens](../guidelines/multi-property.md) but don't have filtering by `EnterpriseIds` or `ChainIds` and will include entities from all enterprises within the scope of the portfolio.
```
Mine suggestion how to write it. It starts as normal + describes that there is no `EnterpriseIds` or `ChainIds` which imho is important because in the description of the Portfolio access token, we're saying there will one of those filters. @MikeAdamsMews will probably even improve the message 😁

Author: Much clearer, thanks!

Reviewer 4: May I ask why it doesn't have EnterpriseIds or ChainIds filtering?

Author: Because we didn't include it in the initial PoC yet.

> If there is already an export with status `Pending` or `Processing` for the specified entity type, request for new export will be rejected. Successful exports remain available for download until the expiration date specified in `ExpiresUtc`.

Reviewer 3: ```suggestion
If there is already an export with status `Pending` or `Processing` for the specified entity type, the request for a new export will be rejected. Successful exports remain available for download until the expiration date specified in `ExpiresUtc`.
```

> If there is already an export with status `Pending` or `Processing` for the specified entity type, request for new export will be rejected.

Reviewer 2: I would probably include this in the first sentance that describes this operation?

```suggestion
Create a pending export unless there is already an export with status `Pending` or `Processing` for the specified entity type in which case the request will be rejected.
```

Reviewer 2: Also I would mention, what happens when there is export that is done and I start a new one - if it's gets deleted or not? 🤔 I can imagine I would be interested in such information as a developer.

Author: I see your point but I don't agree about putting it into the first sentence. In general, I think it's preferred to start with basic situation (= there are 0 existing exports) and then dive into details like this one. I agree about your second comment though.

Author: Added the following sentence:

> Successful exports remain available for download until the expiration date specified in `ExpiresUtc`.
